backend/src/main/java/com/flightsearch/backend/Services/SearchService.java

```diff
@@ -91,15 +108,27 @@ private JSONObject addAirportAndAirlinesCommonNames(String originalJson) throws
                     JSONObject segment = segments.getJSONObject(k);
 
                     String deptIATA = segment.getJSONObject("departure").getString("iataCode");
-                    String deptAirportName = airportCodeService.airportNameSearchByKeyword(deptIATA);
+                    String deptAirportName = deptAirportNamesByIATACode.get(deptIATA);
+                    if (deptAirportName == null) {
+                        deptAirportName = airportCodeService.airportNameSearchByKeyword(deptIATA);
+                        deptAirportNamesByIATACode.put(deptIATA, deptAirportName);
+                    }
                     segment.getJSONObject("departure").put("airportCommonName",deptAirportName);
 
                     String arrIATA = segment.getJSONObject("arrival").getString("iataCode");
-                    String arrAirportName = airportCodeService.airportNameSearchByKeyword(arrIATA);
+                    String arrAirportName = arrAirportNamesByIATACode.get(arrIATA);
+                    if (arrAirportName == null) {
+                        arrAirportName = airportCodeService.airportNameSearchByKeyword(arrIATA);
+                        arrAirportNamesByIATACode.put(arrIATA, arrAirportName);
+                    }
                     segment.getJSONObject("arrival").put("airportCommonName",arrAirportName);
 
                     String carrierCode = segment.getString("carrierCode");
-                    String airlineName = airlineInformationService.airlineNameLookUp(carrierCode);
+                    String airlineName = airlineNamesByCarrierCode.get(carrierCode);
+                    if (airlineName == null) {
+                        airlineName = airlineInformationService.airlineNameLookUp(carrierCode);
+                        airlineNamesByCarrierCode.put(carrierCode, airlineName);
+                    }
                     segment.put("airlineCommonName",airlineName);
                 }
             }
```

Comment:
Comment for `sort` method:

Usually the sort functionality is implemented in the same API call, in this case the /flightOffer endpoint should include a query parameter (e.g. "[sort](https://learn.microsoft.com/en-us/azure/architecture/best-practices/api-design#filter-and-paginate-data)")

> You can use a similar strategy to sort data as it is fetched, by providing a sort parameter that takes a field name as the value, such as /orders?sort=ProductID

With your approach, if the client hasn't called the /flightOffer endpoint, this call won't work since the instance variable "flightOffers" hasn't been initialized yet.

Remember that API calls should be stateless: one call should not depend on another. Unless you're persisting client information in HTTP session for example.